File: bae-android/app/src/main/java/fm/bae/app/playback/PlaybackPositionModel.kt
package fm.bae.app.playback

import fm.bae.app.formatDurationMs
import fm.bae.app.formatRemainingMs

/**
 * Pure seek and position math for the now-playing bar and the Media3 playback
 * state. Owns the position anchor from core's progress events, the current
 * track's duration, the latest raw progress fraction, and the pending in-track
 * seek whose dropped position wins until core's progress catches up.
 *
 * Holds no Media3 or bridge types: track ids are plain strings, positions and
 * durations are milliseconds, and an unknown duration is null. [BaeCorePlayer]
 * holds one instance, feeds it the playback events, and reads its outputs for
 * both the Media3 state and the position StateFlow.
 */
internal class PlaybackPositionModel {
    /** Position anchor from the latest progress event. */
    private var anchorPositionMs: Long = 0L

    /** The current track's duration, or null when core has reported none. Non-null
     *  is always positive (a reported 0 means "unknown" and maps to null). */
    var durationMs: Long? = null
        private set

    /** Latest raw [0,1] progress fraction from core, used for the slider when no
     *  duration is known to derive it from the position. */
    private var progress: Double = 0.0

    private var pendingSeek: PendingSeek? = null

    /** The position to render: the pending seek's target while it holds, else the
     *  last anchor core reported. */
    val effectivePositionMs: Long
        get() = pendingSeek?.targetPositionMs ?: anchorPositionMs

    private data class PendingSeek(
        val trackId: String?,
        val targetPositionMs: Long,
    ) {
        fun matches(trackId: String): Boolean = this.trackId == null || this.trackId == trackId
    }

    /**
     * Fold a progress event into the position state. A pending in-track seek for
     * the incoming track outranks live progress ([PositionUpdate.HeldByPendingSeek])
     * until [onSeeked] confirms it; a position for a different track than the one
     * playing is rejected ([PositionUpdate.StaleTrack]); otherwise the anchor,
     * duration, and progress advance ([PositionUpdate.Applied]).
     */
    fun onProgress(
        currentTrackId: String?,
        trackId: String,
        positionMs: Long,
        durationMs: Long,
        progress: Double,
    ): PositionUpdate {
        val pendingSeek = pendingSeek
        if (pendingSeek != null && pendingSeek.matches(trackId)) {
            return PositionUpdate.HeldByPendingSeek
        }
        return apply(currentTrackId, trackId, positionMs, durationMs, progress)
    }

    /**
     * Fold a seek-confirmation event into the position state, clearing the pending
     * seek it confirms. A position for a different track than the one playing is
     * rejected ([PositionUpdate.StaleTrack]).
     */
    fun onSeeked(
        currentTrackId: String?,
        trackId: String,
        positionMs: Long,
        durationMs: Long,
        progress: Double,
    ): PositionUpdate {
        if (!targetsCurrentTrack(currentTrackId, trackId)) {
            return PositionUpdate.StaleTrack
        }
        pendingSeek = null
        return apply(currentTrackId, trackId, positionMs, durationMs, progress)
    }

    private fun apply(
        currentTrackId: String?,
        trackId: String,
        positionMs: Long,
        durationMs: Long,
        progress: Double,
    ): PositionUpdate {
        if (!targetsCurrentTrack(currentTrackId, trackId)) {
            return PositionUpdate.StaleTrack
        }
        anchorPositionMs = positionMs
        this.durationMs = durationMs.takeIf { it > 0L }
        this.progress = progress
        return PositionUpdate.Applied
    }

    /** A position update targets the current track when no track is playing yet or
     *  it names the one that is; a position for a different track is stale. */
    private fun targetsCurrentTrack(
        currentTrackId: String?,
        trackId: String,
    ): Boolean = currentTrackId == null || currentTrackId == trackId

    /**
     * Enter or refresh the active track. Clears the pending seek when the track
     * changed (its projection no longer applies) and adopts the reported duration
     * ([rawDurationMs] of 0 means unknown). Leaves the anchor and progress to the
     * progress events.
     */
    fun setActiveTrack(
        trackChanged: Boolean,
        rawDurationMs: Long,
    ) {
        if (trackChanged) {
            pendingSeek = null
        }
        durationMs = rawDurationMs.takeIf { it > 0L }
    }

    /**
     * Begin an in-track seek to [requestedPositionMs], projecting the dropped
     * position (clamped into the track) until core's progress confirms it. Returns
     * the [0,1] ratio to send core, or null when there is no known duration to seek
     * within — the caller logs and ignores the seek.
     */
    fun beginInTrackSeek(
        trackId: String?,
        requestedPositionMs: Long,
    ): Double? {
        val total = durationMs ?: return null
        val targetPositionMs = requestedPositionMs.coerceIn(0L, total)
        pendingSeek = PendingSeek(trackId, targetPositionMs)
        return targetPositionMs.toDouble() / total.toDouble()
    }

    /** Drop the pending seek without confirming it — skipping to another queue
     *  entry abandons an in-track seek projected on the current one. */
    fun clearPendingSeek() {
        pendingSeek = null
    }

    /** Reset to the no-playback state. */
    fun reset() {
        pendingSeek = null
        anchorPositionMs = 0L
        durationMs = null
        progress = 0.0
    }

    /**
     * The seek-bar projection: the [0,1] slider fraction plus core's pre-formatted
     * elapsed/remaining labels. When a duration is known the fraction is derived
     * from the position; otherwise core's raw progress is used. Empty when no
     * track plays ([hasCurrentTrack] false).
     */
    fun position(hasCurrentTrack: Boolean): PlaybackPosition {
        if (!hasCurrentTrack) {
            return PlaybackPosition(0.0, "", "")
        }
        val positionMs = effectivePositionMs
        val total = durationMs
        val progress =
            if (total != null) {
                (positionMs.toDouble() / total.toDouble()).coerceIn(0.0, 1.0)
            } else {
                this.progress.coerceIn(0.0, 1.0)
            }
        return PlaybackPosition(
            progress = progress,
            elapsedLabel = formatDurationMs(positionMs),
            remainingLabel = if (total != null) formatRemainingMs(positionMs, total) else "",
        )
    }
}

/** The outcome of folding a progress or seek position into [PlaybackPositionModel]. */
internal enum class PositionUpdate {
    /** The position advanced; the player should republish its projection. */
    Applied,

    /** A pending in-track seek for the current track outranks this live progress
     *  until core confirms the seek — the projection holds, nothing republishes. */
    HeldByPendingSeek,

    /** The position is for a track that is not the one playing. */
    StaleTrack,
}

File: bae-avalonia/Stores/PlaybackPositionModel.cs
using System;

namespace Bae.Desktop;

// The now-playing bar's seek projection: the position the user dropped the
// slider on, shown until core confirms it. TrackId pins it to the track it was
// made for so a projection for a since-changed track is dropped.
public sealed record SeekProjection(
    string? TrackId,
    ulong TargetPositionMs,
    ulong DurationMs,
    double Progress);

public sealed record PlaybackPositionSnapshot(
    ulong DurationMs,
    long PositionMs,
    double Progress);

public sealed record PlaybackPositionState(
    PlaybackPositionSnapshot Snapshot,
    SeekProjection? Projection);

public sealed record NowPlayingState(
    string? AlbumId,
    string? TrackId,
    PlaybackPositionState? Position);

// Whether a playback-position update targets the current track, and if not, why.
public enum PlaybackPositionRejection
{
    Accepted,
    MissingTrackId,
    StaleTrack,
}

// Pure position math for the now-playing bar: seek projection, remaining time,
// and the state transitions the playback store applies. No Avalonia, no bridge —
// track ids are plain strings, slider bounds are doubles.
public static class PlaybackPositionModel
{
    // The seek the user dropped the slider on, as a projection for the current
    // snapshot: clamp the requested progress into the slider's range, map it to a
    // 0..1 ratio, and round to the target position, never past the duration.
    public static SeekProjection ProjectSeek(
        string? trackId,
        PlaybackPositionSnapshot current,
        double requestedProgress,
        double sliderMinimum,
        double sliderMaximum)
    {
        var progress = Math.Clamp(requestedProgress, sliderMinimum, sliderMaximum);
        var progressRange = sliderMaximum - sliderMinimum;
        if (progressRange <= 0)
        {
            throw new InvalidOperationException("Seek slider range must be positive.");
        }

        var ratio = Math.Clamp((progress - sliderMinimum) / progressRange, 0, 1);
        var targetPositionMs = (ulong)Math.Round(current.DurationMs * ratio);
        if (targetPositionMs > current.DurationMs)
        {
            targetPositionMs = current.DurationMs;
        }

        return new SeekProjection(trackId, targetPositionMs, current.DurationMs, progress);
    }

    // A projection wins over a live progress update only when it was made for the
    // same track; a projection for a track that has since changed is dropped.
    public static bool ProjectionWins(SeekProjection? projection, string trackId) =>
        projection is not null && projection.TrackId == trackId;

    // Whether a playback-position update targets the current track: a missing
    // incoming id or an id for a different track than the one now playing is
    // rejected.
    public static PlaybackPositionRejection ClassifyPlaybackPosition(string? currentTrackId, string? incomingTrackId)
    {
        if (incomingTrackId is null)
        {
            return PlaybackPositionRejection.MissingTrackId;
        }
        if (currentTrackId is not null && currentTrackId != incomingTrackId)
        {
            return PlaybackPositionRejection.StaleTrack;
        }
        return PlaybackPositionRejection.Accepted;
    }

    // The leading label's tooltip names the action a click performs, so it follows
    // the current mode.
    public static string TimeLabelTooltipKey(bool showRemaining) =>
        showRemaining ? "nowplaying.show_elapsed" : "nowplaying.show_remaining";

    // Fold a fresh snapshot into the now-playing state, keeping the album when a
    // state already exists, or creating one for the track when it doesn't.
    public static NowPlayingState WithPosition(
        NowPlayingState? current,
        string trackId,
        PlaybackPositionSnapshot snapshot,
        SeekProjection? projection)
    {
        var position = new PlaybackPositionState(snapshot, projection);
        return current is { } nowPlaying
            ? nowPlaying with { Position = position }
            : new NowPlayingState(null, trackId, position);
    }

    // Drop the seek projection while keeping the underlying position snapshot.
    public static NowPlayingState? ClearProjection(NowPlayingState? state) =>
        state is { Position: { } position } nowPlaying
            ? nowPlaying with { Position = position with { Projection = null } }
            : state;

    // Enter a loading transition: reset the position when the loading track
    // differs from the one on screen, keeping it otherwise.
    public static NowPlayingState? BeginLoading(NowPlayingState? current, string trackId) =>
        current is { } state && trackId != state.TrackId
            ? state with { Position = null }
            : current;
}
